safe/src/lib.rs

```diff
@@ -0,0 +1,155 @@
+//! serum-safe defines the interface for the serum safe program.
+
+#![cfg_attr(feature = "strict", deny(warnings))]
+
+use solana_client_gen::prelude::*;
+
+#[cfg_attr(feature = "client", solana_client_gen)]
+pub mod instruction {
+    use super::*;
+    #[derive(serde::Serialize, serde::Deserialize)]
+    pub enum SafeInstruction {
+        /// Initializes a safe instance for use.
+        ///
+        /// Similar to a token mint, this must be included in the same
+        /// instruction that creates the Safe account to initialize. Otherwise
+        /// someone can take control of the account by calling initialize on it.
+        ///
+        /// Accounts:
+        ///
+        /// 0. `[writable]` Safe to initialize.
+        /// 1. `[]`         Mint of the SPL token controlled by the safe.
+        /// 2. `[]`         Rent sysvar
+        Initialize {
+            /// The priviledged account.
+            authority: Pubkey,
+            /// The nonce to use to create the Safe's derived-program address,
+            /// which is used as the authority for the safe's token vault.
+            nonce: u8,
+        },
+        /// Deposit creates deposit and vesting account, transferring tokens
+        /// from the controlling token account to one owned by the SrmSafe
+        /// program. Anyone with funds to deposit can invoke this instruction.
+        ///
+        /// Accounts:
+        ///
+        /// 0. `[writable]  Vesting account representing this deposit.
+        ///                 The owner of this account is this program.
+        ///                 It's data size is dynamic.
+        /// 1. `[writable]` Depositor token account, transferring ownership
+        ///                 *from*, itself to this program.
+        /// 2. `[signer]`   The authority/owner/delegate of Accounts[1].
+        /// 3. `[writable]` The program controlled token vault, transferring
+        ///                 ownership *to*. The owner of this account is the
+        ///                 program derived address with nonce set in the
+        ///                 Initialize instruction.
+        /// 4. `[]`         Safe instance.
+        /// 5. `[]`         SPL token program.
+        /// 6. `[]`         Rent sysvar.
+        #[cfg_attr(feature = "client", create_account(..))]
+        Deposit {
+            /// The beneficiary of the vesting account, i.e.,
+            /// the user who will own the SRM upon vesting.
+            vesting_account_beneficiary: Pubkey,
+            /// The Solana slot number at which point a vesting amount unlocks.
+            vesting_slots: Vec<u64>,
```

Comment:
Will be switching this instruction to use a linear vesting schedule based on an end slot. So these `Vec<>` fields will go away.